mirrord/cli/src/port_forward.rs

```diff
@@ -0,0 +1,689 @@
+                        tracing::trace!("Successful connection to remote address {remote_socket}, connection ID is {}", res.connection_id);
+                    }
+                    Err(error) => {
+                        return Err(PortForwardError::AgentError(format!(
+                            "Problem receiving DaemonTcpOutgoing::Connect {error}"
+                        )))
+                    }
+                },
+                DaemonTcpOutgoing::Read(res) => match res {
+                    Ok(res) => {
+                        let Some(&remote_socket) = self.sockets.get(&res.connection_id) else {
+                            return Err(PortForwardError::SocketFromIdNotFound(res.connection_id));
+                        };
+                        let Some(sender) = self.tasks.get(&remote_socket) else {
+                            return Err(PortForwardError::SenderNotFound(
+                                remote_socket,
+                                res.connection_id,
+                            ));
+                        };
+                        let Ok(_) = sender.send(res.bytes).await else {
+                            return Err(PortForwardError::InternalMessageError(
+                                "Failed to send data from remote to task".into(),
+                            ));
+                        };
+                    }
+                    Err(error) => {
+                        return Err(PortForwardError::AgentError(format!(
+                            "Problem receiving DaemonTcpOutgoing::Read {error}"
+                        )))
+                    }
+                },
+                DaemonTcpOutgoing::Close(connection_id) => {
+                    let Some(remote_socket) = self.sockets.remove(&connection_id) else {
+                        return Err(PortForwardError::SocketFromIdNotFound(connection_id));
+                    };
+                    let Some(sender) = self.tasks.remove(&remote_socket) else {
+                        return Err(PortForwardError::SenderNotFound(
+                            remote_socket,
+                            connection_id,
+                        ));
+                    };
+                    drop(sender);
+                    tracing::trace!(
+                        "Connection closed for remote socket {remote_socket}, connection {connection_id}"
+                    );
+                }
+            },
+            DaemonMessage::LogMessage(log_message) => match log_message.level {
+                LogLevel::Warn => tracing::warn!(log_message.message),
+                LogLevel::Error => tracing::error!(log_message.message),
+            },
+            DaemonMessage::Pong => {
+                if !self.waiting_for_pong {
+                    // message not expected
+                    return Err(PortForwardError::AgentError(
+                        "Unexpected message from Agent: DaemonMessage::Pong".into(),
+                    ));
```

Comment:
I feel like I see a lot of `PortForwardError::AgentError`s with similar message. Maybe you can add `UnexpectedMessage(DaemonMessage)` variant to the error enum?

Reply:
I think the restructuring so far has already eliminated this 👍